keps/sig-scheduling/3094-pod-topology-spread-considering-taints/README.md

```diff
@@ -0,0 +1,824 @@
+know that this has succeeded?
+-->
+
+- Introduce a new struct to define all node inclusion policies explicitly
+- Provide an option for end-users to specify whether to respect taints or not
+
+### Non-Goals
+
+<!--
+What is out of scope for this KEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+- Support customized taints in array
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+Introduce a new field to `TopologySpreadConstraint` to define all node inclusion
+policies including nodeAffinity and nodeTaint.
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+When calculating pod topology spread skew, I want to exclude nodes that don't tolerate
+all taints to prevent pods from falling into unexpected Pending state.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+- Looping all the toleration/taints to filter out unsatisfied nodes may
+lead to performance problem.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+A new field named `NodeInclusionPolicies` will be introduced to `TopologySpreadConstraint`:
+```golang
+type TopologySpreadConstraint struct {
+  // NodeInclusionPolicies includes several policies
+  // when calculating pod topology spread skew
+  NodeInclusionPolicies nodeInclusionPolicies
+}
+```
+
+There are two policies now regarding to nodeAffinity and nodeTaint:
+```golang
+type nodeInclusionPolicies struct {
+  // Respect nodeAffinity/nodeSelector or not in calculating.
```

Comment:
This requires better wording: calculating what? Maybe something like:

Policy for the Pod's node affinity and node selectors when calculating matching pods. The values could be:
- Respect: ....
- Ignore: ....
By default, the policy is Respect.